File: sdk/apps/examples/hooks/PreToolUse.ts
#!/usr/bin/env bun
/**
 * Cline Hook: PreToolUse (TypeScript)
 * Logs and filters tool calls with type safety.
 * Copy to ~/.cline/hooks/PreToolUse.ts and chmod +x
 */

interface ToolCall {
	id: string;
	name: string;
	input: Record<string, unknown>;
}

interface HookEvent {
	hookName: string;
	tool_call: ToolCall;
	preToolUse: {
		toolName: string;
		parameters: Record<string, unknown>;
	};
}

interface HookControl {
	cancel?: boolean;
	review?: boolean;
	context?: string;
	errorMessage?: string;
	overrideInput?: unknown;
}

async function main(): Promise<void> {
	let event: HookEvent;

	try {
		const input = await Bun.stdin.text();
		event = JSON.parse(input) as HookEvent;
	} catch (error) {
		const message =
			error instanceof Error ? error.message : "Unknown parse error";
		console.log(
			JSON.stringify({ errorMessage: `Failed to parse: ${message}` }),
		);
		return;
	}

	const toolName = event.tool_call.name;
	const parameters = event.preToolUse.parameters;

	// Log to stderr
	console.error(`🔧 Tool: ${toolName}`);
	if (Object.keys(parameters).length > 0) {
		const paramStr = Object.entries(parameters)
			.map(([k, v]) => `${k}=${v}`)
			.join(", ");
		console.error(`   Args: ${paramStr}`);
	}

	// Return empty control (allow execution)
	const control: HookControl = {};
	console.log(JSON.stringify(control));
}

main().catch((error) => {
	console.error("Hook error:", error);
	process.exit(1);
});

File: sdk/examples/hooks/PreToolUse.py
#!/usr/bin/env python3
"""
Cline Hook: PreToolUse (Python)
Logs every tool call before it executes.
Copy to ~/.cline/hooks/PreToolUse.py and chmod +x
"""

import sys
import json

def main():
    # Read event payload from stdin
    try:
        event = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        print(json.dumps({"errorMessage": f"Failed to parse hook input: {e}"}))
        return

    # Extract tool information
    tool_call = event.get("tool_call", {})
    tool_name = tool_call.get("name", "unknown")
    parameters = event.get("preToolUse", {}).get("parameters", {})

    # Log to stderr
    print(f"🔧 Tool: {tool_name}", file=sys.stderr)
    if parameters:
        param_str = ", ".join(f"{k}={v}" for k, v in parameters.items())
        print(f"   Args: {param_str}", file=sys.stderr)

    # Return empty control object (allow the tool to execute)
    print(json.dumps({}))

if __name__ == "__main__":
    main()
